firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java

```diff
@@ -455,59 +468,83 @@ public void beginRealtimeHttpStream() {
-      responseCode = httpURLConnection.getResponseCode();
-
-      // If the connection returned a 200 response code, start listening for messages.
-      if (responseCode == HttpURLConnection.HTTP_OK) {
-        // Reset the retries remaining if we opened the connection without an exception.
-        resetRetryCount();
-        metadataClient.resetRealtimeBackoff();
-
-        // Start listening for realtime notifications.
-        ConfigAutoFetch configAutoFetch = startAutoFetch(httpURLConnection);
-        configAutoFetch.listenForNotifications();
-      }
-    } catch (IOException e) {
-      // Stream could not be open due to a transient issue and the system will retry the connection
-      // without user intervention.
-      Log.d(TAG, "Exception connecting to real-time RC backend. Retrying the connection...", e);
-    } finally {
-      closeRealtimeHttpStream(httpURLConnection);
-      setIsHttpConnectionRunning(false);
-
-      boolean connectionFailed = responseCode == null || isStatusCodeRetryable(responseCode);
-      if (connectionFailed) {
-        updateBackoffMetadataWithLastFailedStreamConnectionTime(
-            new Date(clock.currentTimeMillis()));
-      }
-
-      // If responseCode is null then no connection was made to server and the SDK should still
-      // retry.
-      if (connectionFailed || responseCode == HttpURLConnection.HTTP_OK) {
-        retryHttpConnectionWhenBackoffEnds();
-      } else {
-        String errorMessage =
-            String.format(
-                "Unable to connect to the server. Try again in a few minutes. HTTP status code: %d",
-                responseCode);
-        // Return server message for when the Firebase Remote Config Realtime API is disabled and
-        // the server returns a 403
-        if (responseCode == 403) {
-          errorMessage = parseForbiddenErrorResponseMessage(httpURLConnection.getErrorStream());
-        }
-        propagateErrors(
-            new FirebaseRemoteConfigServerException(
-                responseCode,
-                errorMessage,
-                FirebaseRemoteConfigException.Code.CONFIG_UPDATE_STREAM_ERROR));
-      }
-    }
+    // Make async call to create and setup HTTP connection.
+    Task<HttpURLConnection> httpURLConnectionTask = createRealtimeConnection();
+    // When the connection task has finished, begin real-time actions.
+    Tasks.whenAllComplete(httpURLConnectionTask)
+        .continueWith(
+            this.scheduledExecutorService,
+            (completedHttpUrlConnectionTask) -> {
+              Integer responseCode = null;
+              HttpURLConnection httpURLConnection = null;
+              setIsHttpConnectionRunning(true);
+
+              try {
+                // If HTTP connection task failed throw exception to move to the catch block.
+                if (!httpURLConnectionTask.isSuccessful()) {
+                  throw new IOException(httpURLConnectionTask.getException());
```

Comment:
Thinking: More refactoring would propagate the failed task, but we're limiting the scope of this change. It's a bit confusing to throw the `IOException` here just to handle it a few lines below.

Reply:
Yeah I think more refactoring can be done to handle the failed task better, I threw the `IOException` here b/c I wanted to keep the same try/catch structure that we had here and to reduce the scope of these changes